File: src/cpp/graph/datastruct/graph.cpp
// %title
// 無向グラフ
// 
// %overview
// 隣接頂点の情報を持つ．
// 必要な空間計算量は |V|+2|E|
//
// %usage
// void connect(int from, int to)
// つなげる．ループ辺は非推奨．
// void resize(size_t _n)
// 頂点数を変更する．
// global変数として宣言した時に用いる．
//
// %verified
// 
// %references
//



class Graph{
public:
    size_t n;
    vector<vector<int>> vertex_to;
    
    Graph(size_t n=1):n(n),vertex_to(n){}
    
    void connect(int from, int to){
        vertex_to[(size_t)from].emplace_back(to);
        vertex_to[(size_t)to].emplace_back(from);
    }
    void resize(size_t _n){
        n = _n;
        vertex_to.resize(_n);
    }
};

File: src/rust/graph/graph.rs
// %=BEGIN DOC
// %title
// グラフ（隣接頂点）
//
// %overview
// 無向・有向どちらも表現可能。
// 辺に情報を持つことができる。
// ただし、無向辺を双方向の有向辺と表現するため、
// 無向辺に情報をもたせる場合は倍のリソースを消費する
// %usage
//
// %require
// ```
// ```
// %verified
// rand
//
// %references
//
// %=END DOC
// %=BEGIN CODE
struct Graph<T: Clone = ()>(Vec<Vec<(usize, T)>>);

impl<T: Clone> Graph<T> {
    fn to<'a>(&self, i: usize) -> &Vec<(usize, T)> {
        &(self.0)[i]
    }
}

struct GraphBuilder<T: Clone = ()>(Graph<T>);

impl<T: Clone> GraphBuilder<T> {
    fn new(n: usize) -> Self {
        Self(Graph::<T>(vec![Vec::<(usize, T)>::new(); n]))
    }
    fn connect(&mut self, u: usize, v: usize, item: T) {
        (self.0 .0)[u].push((v, item.clone()));
        (self.0 .0)[v].push((u, item));
    }
    fn connect_d(&mut self, src: usize, dst: usize, item: T) {
        (self.0 .0)[src].push((dst, item));
    }
    fn build(self) -> Graph<T> {
        self.0
    }
}
// %=END CODE
